Select folder at save bar

dm3ch opened the pull request:
I'm trying to implement https://community.bitwarden.com/t/add-collection-selection-to-save-password-popup-in-browser/14643 myself.

I am trying to add selecting of folder for saving login in notification bar.
I have found how to pass additional data from ts code to notification bar.
But for now I don't understand how to get access to FolderService from notificationBar.ts

I would be happy if anyone could answer me how to do that.

P.S. I'm newbie in frontend

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/bitwarden/browser?pullRequest=1409) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/bitwarden/browser?pullRequest=1409) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/bitwarden/browser?pullRequest=1409) it.</sub>

dm3ch commented:
@kspearrin Could you please point me right direction?

cscharf commented:
Hi @dm3ch , re:

> But for now I don't understand how to get access to FolderService from notificationBar.ts

This type of component is not intended to get dependencies such as those services passed to it or to be used directly. It's purely intended for background processes/signals to be passed, loaded, etc. to the rest of the DOM or event listeners that can respond to those events from the popup. I imagine it would look something like what you see in the event pair of `bgGetDataForTab` and `message` (of type, `notificationBarDataResponse`).

I'll leave it to @kspearrin  with any other info or detail. Please be patient as it may take Kyle a little bit to respond.

kspearrin commented:
You cannot access services from the content scripts. The only way to get data there is to pass it via message.

dm3ch commented:
Ok, got it, thank you guys, i'll try to implement that this way.

dm3ch commented:
So, i finished selecting folder implementation in notification bar.

I tested that on firefox, and that worked for me.
But I have some problems with implementing default selection of 'No folder'.

Please review.

dm3ch commented:
@cscharf Sorry, for mention, but does it need something additional to get review? Cause bitwarden contributing guide contains no guide for PRs.

cscharf commented:
Hi @dm3ch , no worries on the mention, the PR overall looks good, I know we don't really have instructions on PR submission guidelines (yet), our only constraint at the moment is time, we want to be able to give this the necessary attention to detail you deserve so we can provide appropriate feedback and help make this something we merge in.

The only thing I would ask is perhaps screen shots of the changes along with some legwork on testing (and screenshots) in Chrome and Safari.

If you're still stuck on the "No Folder" default selection let us know as well so perhaps someone from the community or one of our devs can take a look at that specifically during the review.

dm3ch commented:
> Hi @dm3ch , no worries on the mention, the PR overall looks good, I know we don't really have instructions on PR submission guidelines (yet), our only constraint at the moment is time, we want to be able to give this the necessary attention to detail you deserve so we can provide appropriate feedback and help make this something we merge in.
> 
> The only thing I would ask is perhaps screen shots of the changes along with some legwork on testing (and screenshots) in Chrome and Safari.
> 
> If you're still stuck on the "No Folder" default selection let us know as well so perhaps someone from the community or one of our devs can take a look at that specifically during the review.

Thank you for answer.

I will create screenshot of changes on Chrome and Safari and manual test cases that i checked.
Suddenly I haven't yet found a guide how to enabled dev extension in Safari...

About problems that is not solved in PR yet:
1) 'No Folder' default option selection, seems to work not so good.
I understand that I can add it to html source, cause this folder is always there, but I would like to set default value from JS, cause I would like to add default folder configuration inside browser extensions settings. So it could be useful in future to set default selection from JS, not HTML.

So I would appreciate any help with that.
2) UI - I'm not sure if we should add comment near selection dropdown or should we add some styles to select element.
So feedback is welcome.

cscharf commented:
> Suddenly I haven't yet found a guide how to enabled dev extension in Safari...

This is a fun topic for sure... You will need to uninstall the desktop Bitwarden instance, if installed. You must have Xcode installed and any other necessary dev profiles, etc. and then you can debug the xcode project directly, it will add-in the extension to Safari if you build for that profile and attach to it... OR, I can build/package it for you from your PR branch when you're ready for testing (probably easier). Just let me know.

dm3ch commented:
So here is:
1) Safari:
<img width="1440" alt="Screenshot 2020-10-02 at 16 52 42" src="https://user-images.githubusercontent.com/5025313/94931715-c2a16d80-04d0-11eb-90e5-c8ab8441da6e.png">

2) Chrome
<img width="1440" alt="Screenshot 2020-10-02 at 16 58 40" src="https://user-images.githubusercontent.com/5025313/94931721-c3d29a80-04d0-11eb-9073-cc9e3b8f1bb8.png">

As you can see styles are different. :(

What have i already tested manually:
* When you select folder, item is saved in selected folder and you could be found it via web Vault.
* When you select 'No Folder', item is saved without folder and could be found in web Vault. Separate test case cause 'No Folder' is corner case when we have got 'null' folder id and we should transform 'null' string to actual null.

These things was tested by me only on Firefox for now, but I assume that on chrome it should also works fine, cause it implements same APIs, but I would test that

dm3ch commented:
Just tested this two scenarios on Chrome all is fine.

@cscharf Could you provide a Safari extension artifact and maybe a link to some blogpost how to install that.

dm3ch commented:
> > Suddenly I haven't yet found a guide how to enabled dev extension in Safari...
> 
> This is a fun topic for sure... You will need to uninstall the desktop Bitwarden instance, if installed. You must have Xcode installed and any other necessary dev profiles, etc. and then you can debug the xcode project directly, it will add-in the extension to Safari if you build for that profile and attach to it... OR, I can build/package it for you from your PR branch when you're ready for testing (probably easier). Just let me know.

And why desktop app should be uninstalled for testing browser extension in safari? it sounds a little bit weird. :)

cscharf commented:
> And why desktop app should be uninstalled for testing browser extension in safari? it sounds a little bit weird. :)

lol, yes, it is odd, however when debugging in XCode with the App Extension, if you have the Bitwarden app installed (official from app store or dmg), it is what contains and packages the Safari extension, you'll get a conflict and not be able to debug your local dev version of it. Essentially you uninstall it and run the project in Xcode for debugging against Safari with your Apple Developer ID account and provisioning profile (which allows you to debug/side-load in Safari), sometimes you have to open Safari a couple of times and open the Settings -> Extensions to ensure it's cleared out and then re-enabled (although new version of Xcode should enable it automatically when you start debugging). This is just the "fun" that is Safari app extensions (so thankful Safari 14 supports web extensions... again)

cscharf commented:
> Could you provide a Safari extension artifact and maybe a link to some blogpost how to install that.

@dm3ch , Yes, let me pull your branch and I can do that, it will essentially be a local/dev build of the Desktop app as a DMG so you'd install it just like any other, it'll also be signed and notarized by Apple allowing you to install it and it to work appropriately.

dm3ch commented:
Apple... their so developer unfriendly... 

Just found their official guide how to load safari extension locally. All browsers allow it to be done by one button, but apple is just apple. :D

cscharf commented on src/notification/bar.js:
```diff
@@ -137,4 +146,33 @@ document.addEventListener('DOMContentLoaded', () => {
+                }
+            }, false);
+        } else {
+            chrome.runtime.onMessage.addListener( (request, sender, response) =>{
```
Minor formatting nit:
```JavaScript
.addListener( (request, sender, response) =>{
// should be
.addListener((request, sender, response) => {
```

cscharf commented on src/notification/bar.js:
```diff
@@ -1,21 +1,20 @@
     var lang = window.navigator.language;
     if (typeof safari !== 'undefined') {
-        const responseCommand = 'notificationBarFrameDataResponse';
+        const responseI18nCommand = 'notificationBarFrameDataResponse';
```
Why did you rename this variable?

cscharf replied:
It's okay, just was curious as to why the rename

cscharf commented on src/notification/bar.js:
```diff
@@ -137,4 +146,33 @@ document.addEventListener('DOMContentLoaded', () => {
+        select = document.getElementById("select-folder");
+        folders.forEach(folder => {
+            //Select "No Folder" folder by default
+            let selected = folder.id === 'null';
```
Is this value coming as a string, `'null'` from the API directly or the folder service? Trying to figure out why it wouldn't just be `''` or `null` explicitly. Not a huge fan of a string value of `'null'` vs. `''` or `null` if we can help it.

cscharf replied:
Is this also where your default select for "No Folder" is coming in? Is the selected attribute of the `Option` not working in this case to default-select it in the notification bar?

dm3ch replied:
So about this:
* Yes this is a place where i'm trying to set default option.
* I'm getting 'null' as a string from folderService, as far as i understands. So I'm just getting map from folderService and return it without any modifications. So I assume that maybe it's json serialisation issue or issue in folder service.  here's that line https://github.com/bitwarden/browser/pull/1409/files#diff-251eb20a94f368cedaceca1d703e4aeeR463

But suddenly I haven't yet found implementation of getAllDecrypted. :(

cscharf replied:
`getAllDecrypted()` is here: [jslib -> folder.service.ts; L73](https://github.com/bitwarden/jslib/blob/master/src/services/folder.service.ts#L73).

The `id` property is being set to a literal `null`, so it could be a binding issue with the `select`. My assumption would be you would get back `null` from `folder.id` here, but unless you bind it to the `Option`'s value as `''` you would get `'null'` as the `<option/>` value because it's JavaScript and why not, lol. So I would check for `null` instead of `'null'` and that should properly detect the no folder option, then when you're binding the `new Option(...)` pass `folder.id || ''` instead of just `folder.id` and that should handle the case where `folder.id` is null and replace it with a more HTML friendly blank string for the value (equivalent of `null`).

dm3ch replied:
@cscharf Thank you. Just fixed as you wrote.

I realised that null to 'null' conversion seems to happen on id insert, so folder.service is not the one who provides null literally, it seems.

Also it fixed default option handling, but for now i don't understand, why 'null' was the problem for that.

cscharf commented:
@dm3ch , glad you were able to get it working, seems to be okay so far from what I've seen and I'll let @kspearrin perhaps take a look if/as he has time as well. I'm looking at the default selection from JS issue you were having as well now (forgot to check during the review)

dm3ch reviewed (commented):
@cscharf Fixed linting issues. And answered about renaming and null as a string.

Also what is you are using for linting? I would like to use it to prevent future styling issues.

cscharf commented:
> Also what is you are using for linting? I would like to use it to prevent future styling issues.

you can simply run `npm run lint`, that has a package script build in for running tslint.

kspearrin commented on src/notification/bar.js:
```diff
@@ -137,4 +146,33 @@ document.addEventListener('DOMContentLoaded', () => {
+            //Select "No Folder" folder by default
+            let selected = folder.id === 'null';
+
+            let opt = new Option(folder.name, folder.id, selected);
```
const

kspearrin commented on src/notification/bar.js:
```diff
@@ -137,4 +146,33 @@ document.addEventListener('DOMContentLoaded', () => {
+
+    function fillSelectorWithFolders(folders) {
+        select = document.getElementById("select-folder");
+        folders.forEach(folder => {
```
parens around folder

dm3ch replied:
Done

kspearrin commented on src/notification/bar.js:
```diff
@@ -137,4 +146,33 @@ document.addEventListener('DOMContentLoaded', () => {
+        select = document.getElementById("select-folder");
+        folders.forEach(folder => {
+            //Select "No Folder" folder by default
+            let selected = folder.id === 'null';
```
const

kspearrin commented on src/notification/bar.js:
```diff
@@ -32,6 +31,15 @@ document.addEventListener('DOMContentLoaded', () => {
+        command: 'bgGetDataForTab',
+        responseCommand: responseFoldersCommand
+    });
+    addPlatformEventListener(responseFoldersCommand, (msg) => {
```
I would wire up the event listener before sending the message

dm3ch replied:
Fixed that, also done this for i18n same safari code.

kspearrin reviewed (changes requested):
We also need to look at how this UI behaves on certain screen sizes. This bar is limited in space and there is already code in here that makes this bar work for various screen and text length scenarios for different languages. This was alreayd an issue in the past and now we are adding even more horizontal content. We may need to hide the fiolder selection altogether for certain sizes, for example.

dm3ch commented on src/notification/bar.js:
```diff
@@ -1,21 +1,20 @@
+const { CommonModule } = require('@angular/common');
```
Not needed, deleted

dm3ch commented on src/background/runtime.background.ts:
```diff
@@ -18,6 +18,7 @@ import { StorageService } from 'jslib/abstractions/storage.service';
 import { SyncService } from 'jslib/abstractions/sync.service';
 import { SystemService } from 'jslib/abstractions/system.service';
 import { VaultTimeoutService } from 'jslib/abstractions/vaultTimeout.service';
+import { FolderService } from 'jslib/abstractions';
```
Fixed

dm3ch commented:
@kspearrin @cscharf Thank you for review.
Just made all requested changes.

@kspearrin mentioned about testing with different screen width, but i'm not sure how to test that properly, cause I don't know which screen width is minimal supported now and on which language text is the longest one.

cscharf commented:
> Kyle mentioned about testing with different screen width, but i'm not sure how to test that properly, cause I don't know which screen width is minimal supported now and on which language text is the longest one.

We use bootstrap UI for responsive design, so I would test using the standard breakpoints for small devices: https://getbootstrap.com/docs/4.0/layout/overview/#responsive-breakpoints and see if that helps. If it essentially gets hidden after size is reduced past the minimum "working" breakpoint, I would simply go with that using the built-in bootstrap UI breakpoint/size visibility classes/helpers.

dm3ch commented:
@cscharf Just found some time to proceed with this PR.

Here's screenshots how it's rendered with different screen width.
![Screenshot 2021-02-17 at 17 41 29](https://user-images.githubusercontent.com/5025313/108220258-99b00b00-7147-11eb-8626-5aa191d193e1.png)
![Screenshot 2021-02-17 at 17 41 38](https://user-images.githubusercontent.com/5025313/108220263-9a48a180-7147-11eb-8d73-dfdd5979e90e.png)
![Screenshot 2021-02-17 at 17 42 17](https://user-images.githubusercontent.com/5025313/108220268-9ae13800-7147-11eb-83f7-3149c3ca622a.png)

cscharf commented:
I've included some folks on this review for UI/UX feedback and assistance thereto, I am by no means an UI'ologist.

JakubJeziorski commented:
any update on when this will be added?

eliykat commented:
Hey @JakubJeziorski, we're currently in the middle of a big release and we'll be evaluating a number of community PRs that we've received soon. I'll set aside some time to review this next week so that we can work through any changes required and keep it moving.

SZE2 commented:
Hi,
Would it be possible to add the ability to Rename the item before saving as well?
By default, the name is the URL, but when the URL isn't a good name, or when there are multiple logins for the same URL it is very helpful to be able to rename them on the spot without going to rename it after.
Thanks for considering it!

JakubJeziorski commented:
Could you consider using <datalist> (dropdown with input field)
This would allow searching and you could also add the ability to create new folders.

https://stackoverflow.com/questions/5650457/html-select-form-with-option-to-enter-custom-value?noredirect=1&lq=1
https://stackoverflow.com/questions/48722834/how-to-use-html-datalist-with-bootstrap

eliykat reviewed (commented):
I've looked over the code and have suggested a few changes. Overall it looks like you've understood the patterns/architecture you need to use, good job.

Other comments:
* elements appear differently in different browsers, I don't think that's a problem
* I wasn't satisfied with the layout - currently the folder dropdown is floating to the left of the buttons without any label or other context. I suggest putting it next to the Save button (because they're related) and renaming "No folder" to "Select folder". "Select folder" can still be treated as "no folder" in the code (i.e. null value), but it provides instruction to the user. Mockup:
![Screen Shot 2021-03-19 at 1 54 05 pm](https://user-images.githubusercontent.com/31796059/111729656-3f77a680-88bb-11eb-9295-752441318740.png)
Make sure to add the UI text in `src/_locales/en/messages.json` and then use the `i18n` service to use it - plenty of examples in the code. (I suggest you do this work in `runtime.background.ts` so that you have access to all the Angular services)
* @cscharf Correct me if I'm wrong but I can't see a bootstrap dependency in our browser client, and in any case, I don't think it would be available in the notificationBar. @dm3ch this means that you'll need to use media queries in `bar.scss`. See [here](https://developer.mozilla.org/en-US/docs/Learn/CSS/CSS_layout/Media_queries) for a general introduction to media queries or [here](https://stackoverflow.com/questions/32484618/how-to-hide-div-for-specific-height-or-width-using-css) for an example StackOverflow answer. The "standard" breakpoints that Chad was referring to are 576px, 768px, 992px and 1200px wide. I suggest you test the page at each of those widths, figure out at which breakpoint the dropdown needs to be hidden, and then use that in the CSS. (frontend aint easy!)
* @SZE2 - we have limited space in the notificationBar and I don't want this PR to go out of scope - let's leave it as is for now
* @JakubJeziorski - I don't think we use datalists elsewhere in our UI, given that this is only a minor improvement, I would like to keep it consistent with the rest of the app

dm3ch commented on src/notification/bar.js:
```diff
@@ -120,4 +130,31 @@ document.addEventListener('DOMContentLoaded', () => {
     }
+
+    function addPlatformEventListener(command, func) {
+        if (typeof safari !== 'undefined') {
```
Moved to chrome.runtime

dm3ch commented:
@eliykat I'm sorry for 3 month delay, I was a little bit busy.

I just have:
* Applied review suggestions
* Merged latest master
* Added check that folder with givenID exists
* Added 'Select folder...' default option with i18n message
* Checked that all still works fine locally

Could you please resume/restart review process? :D

dm3ch commented:
@eliykat Friendly reminder. Any chance to get new round of review?

cscharf commented:
Hi @dm3ch , Thomas is on vacation at the moment and should be back in about a week and has a few items to tackle but will definitely circle around and review the updates. Thanks!

dm3ch commented:
> Hi @dm3ch , Thomas is on vacation at the moment and should be back in about a week and has a few items to tackle but will definitely circle around and review the updates. Thanks!

Thanks for answer

eliykat commented:
@dm3ch I'm back on deck, I'll check it out this week :)

eliykat commented on src/background/runtime.background.ts:
```diff
@@ -253,6 +255,12 @@ export default class RuntimeBackground {
 
+            const folders = await this.folderService.getAllDecrypted();
+            const folderExist = folders.some((x) => x.id === folderId);
+            if (folderExist && !Utils.isNullOrWhitespace(folderId)) {
```
The `isNullOrWhitespace` check should be done at the start of this block. If `true`, then you can just skip this entire block.

In other words, if the user hasn't selected a folder, then there's no need to get the decrypted folders and look for an `id` match.

dm3ch replied:
Refactored this piece of code

eliykat commented on src/notification/bar.scss:
```diff
@@ -96,3 +96,9 @@ body[class*='lang-en'] .add-buttons {
+
+@media (max-width: 768px) {
+    #select-folder{
+        display: none !important; 
```
Is the `!important` property required? It's usually bad practice to use it, and I assume it's probably not needed here because using an `id` as the selector will take precedence over most other selectors.

dm3ch replied:
Removed !important

eliykat commented on src/notification/bar.js:
```diff
@@ -19,6 +19,18 @@ document.addEventListener('DOMContentLoaded', () => {
     setTimeout(load, 50);
     
 
+    const responseFoldersCommand = 'notificationBarGetFoldersList';
```
We only need this data if we're adding a new login, but currently we're fetching it every time the notificationBar opens. I recommend putting this block in a function, and then calling that function below in the ` if (getQueryVariable('add'))` block.

That also prevents us from trying to populate the folder list in cases where it's not rendered.

dm3ch replied:
Done

eliykat commented on src/notification/bar.html:
```diff
@@ -27,6 +27,9 @@
             <tbody>
                 <tr>
                     <td class="add-text"></td>
+                    <td align="right" class="add-folder-text">
```
This class name doesn't reflect the content (there's no `text` in this table cell). Unless it's actually needed somewhere, I think you can just remove this `class` attribute.

dm3ch replied:
Removed class

eliykat reviewed (changes requested):
Hi @dm3ch, I'm glad you came back to this! It's looking good, especially the responsive elements. I've requested a few changes below, and I'm also going to get our design team to have a quick look at the layout to make sure they're happy. 🚀

dm3ch commented on src/background/runtime.background.ts:
```diff
@@ -21,7 +22,7 @@ import BrowserPlatformUtilsService from '../services/browserPlatformUtils.servic
 
 import MainBackground from './main.background';
-
+ 
```
Fixed

dm3ch commented on src/background/runtime.background.ts:
```diff
@@ -253,6 +255,12 @@ export default class RuntimeBackground {
             model.login = loginModel;
 
+            const folders = await this.folderService.getAllDecrypted();
+            const folderExist = folders.some((x) => x.id === folderId);
```
Refactored this piece of code

dm3ch commented:
> Hi @dm3ch, I'm glad you came back to this! It's looking good, especially the responsive elements. I've requested a few changes below, and I'm also going to get our design team to have a quick look at the layout to make sure they're happy. 🚀

JFYI: Your review comments have been resolved

dm3ch commented:
@eliykat Does any actions needed from my side or we are just waiting design team review?

eliykat commented:
Thanks for following up. The design feedback is just to group the "Select folder" dropdown and the "Save" button together, and separate them from the "Never for this website" button, like in my previous mockup:
![Screen Shot 2021-03-19 at 1 54 05 pm](https://user-images.githubusercontent.com/31796059/111729656-3f77a680-88bb-11eb-9295-752441318740.png)

I'd recommend to keep the spacing between "Never" and the "select folder" option fixed as the browser scales.

dm3ch commented:
@eliykat I had take a look how to make elements order same as on mackup above and that it's not so easy as I thought.

Right now "Yes, Save Now" and "Never for this website" buttons both lives in same `td` - https://github.com/bitwarden/browser/blob/master/src/notification/bar.html#L30-L33

And that `td` have pre-calculated width - https://github.com/bitwarden/browser/blob/master/src/notification/bar.scss#L84-L86

So I'm not sure how to perform proposed change and not to brake anything. :(

eliykat commented:
I'll have a look at it tomorrow and see if I can figure it out, that way we can get it merged for the upcoming release.

eliykat commented:
@dm3ch, this is now done. It was tricky because we don't have a properly responsive design in the notificationBar - some layout is set on load only (e.g. which text we use on the buttons), and other parts are changed dynamically using media queries. This is bad - we should be using media queries all the way through - but for now I've just stuck to the current patterns. But just so you know, what I've done is not a good example. 😆 

I also did some minor tweaks and code cleanup. Thanks for your work on this!

dm3ch commented:
@eliykat Thank you for helping make this PR merged. I appreciate it.

z-l-p commented:
Sorry to comment on an already-merged feature, but I noticed the drop-down only surfaces personal folders, not organizations/collections. Am I missing something or is this the case?

eliykat commented:
@z-l-p that's correct, collections are not currently supported from the save bar. One of our team members is looking into adding this, but no ETA yet.
